File: packages/template-go/parser/blocktag.go
package parser

import (
	"fmt"
	"regexp"
	"strings"

	"github.com/polyspec/template/ast"
	"github.com/polyspec/template/errs"
	"github.com/polyspec/template/expr"
	"github.com/polyspec/template/lexer"
)

var identPattern = regexp.MustCompile(`^[A-Za-z_][A-Za-z0-9_]*$`)

func isPathChar(c byte) bool {
	return c >= 'A' && c <= 'Z' || c >= 'a' && c <= 'z' || c >= '0' && c <= '9' || c == '_' || c == '.' || c == '/' || c == '-'
}

type rawToken struct {
	kind  string // "path" or "ident"
	value string
	start int
	end   int
}

// rawTagReader reads path tokens (GRM-3) and block tag bodies (GRM-13 to GRM-15).
type rawTagReader struct {
	source     *lexer.Source
	close      string
	closeCount int
}

func (r *rawTagReader) text() string { return r.source.Text }

func (r *rawTagReader) atClose(index int) bool {
	return strings.HasPrefix(r.text()[index:], strings.Repeat(r.close, r.closeCount))
}

func (r *rawTagReader) fail(code errs.Code, start, end int, message string) error {
	return r.source.Error(code, start, end, message)
}

// readToken returns nil at the close delimiter or the end of input.
func (r *rawTagReader) readToken(index int) (*rawToken, error) {
	text := r.text()
	index = skipHorizontalSpace(text, index)
	if index >= len(text) || r.atClose(index) {
		return nil, nil
	}
	c := text[index]
	if c == '"' || c == '\'' {
		literal, err := expr.LexStringLiteral(r.source, index)
		if err != nil {
			return nil, err
		}
		return &rawToken{kind: "path", value: literal.Decoded, start: index, end: literal.End}, nil
	}
	end := index
	for end < len(text) && isPathChar(text[end]) {
		end++
	}
	value := text[index:end]
	if strings.ContainsAny(value, "./") {
		return &rawToken{kind: "path", value: value, start: index, end: end}, nil
	}
	if identPattern.MatchString(value) {
		return &rawToken{kind: "ident", value: value, start: index, end: end}, nil
	}
	return &rawToken{kind: "ident", value: "", start: index, end: index + 1}, nil
}

// readIncludePath implements GRM-12.
func (r *rawTagReader) readIncludePath(index int) (string, int, error) {
	token, err := r.readToken(index)
	if err != nil {
		return "", 0, err
	}
	if token == nil || token.value == "" {
		at := skipHorizontalSpace(r.text(), index)
		return "", 0, r.fail(errs.ParseInvalidPath, at, at+1, "include requires a path")
	}
	if token.kind != "path" {
		return "", 0, r.fail(errs.ParseInvalidPath, token.start, token.end, fmt.Sprintf("%q is not a path", token.value))
	}
	return token.value, token.end, nil
}

type blockBody struct {
	id    *string
	path  *string
	scope []*ast.ScopeItem
	end   int
}

// readBlockBody implements GRM-13 to GRM-15.
func (r *rawTagReader) readBlockBody(index int) (*blockBody, error) {
	body := &blockBody{scope: []*ast.ScopeItem{}}
	cursor := index
	token, err := r.readToken(cursor)
	if err != nil {
		return nil, err
	}
	if token == nil || token.value == "" {
		at := skipHorizontalSpace(r.text(), cursor)
		return nil, r.fail(errs.ParseInvalidBlockTag, at, at+1, "block tag requires an identifier or a path")
	}
	if token.kind == "path" {
		v := token.value
		body.path = &v
		cursor = token.end
	} else {
		v := token.value
		body.id = &v
		cursor = token.end
		token, err = r.readToken(cursor)
		if err != nil {
			return nil, err
		}
		if token != nil && token.kind == "path" {
			p := token.value
			body.path = &p
			cursor = token.end
		}
	}
	for {
		token, err = r.readToken(cursor)
		if err != nil {
			return nil, err
		}
		if token == nil {
			break
		}
		if token.kind != "ident" || token.value == "" {
			return nil, r.fail(errs.ParseInvalidBlockTag, token.start, token.end, fmt.Sprintf("unexpected %q in block tag", r.text()[token.start:token.end]))
		}
		cursor = token.end
		text := r.text()
		if cursor < len(text) && text[cursor] == ':' {
			valueStart := cursor + 1
			if valueStart >= len(text) || isHorizontalSpace(text[valueStart]) {
				return nil, r.fail(errs.ParseInvalidBlockTag, cursor, cursor+1, "scope item requires a value after \":\"")
			}
			parser := expr.NewParser(r.source, valueStart, expr.Options{Close: r.close, CloseCount: r.closeCount, OpenIndex: -1})
			value, err := parser.ParsePostfix(true)
			if err != nil {
				return nil, err
			}
			body.scope = append(body.scope, &ast.ScopeItem{Name: token.value, Expr: value})
			cursor = parser.End()
		} else {
			body.scope = append(body.scope, &ast.ScopeItem{Name: token.value, Expr: &ast.Var{Type: "Var", Name: token.value, Span: ast.Span{token.start, token.end}}})
		}
	}
	body.end = cursor
	return body, nil
}

File: packages/template-ts/src/parser/block-tag.ts
// Path tokens (GRM-3) and block tag bodies (GRM-13 to GRM-15).
import type { Expr, ScopeItem } from '../ast.js';
import type { TemplateError } from '../errors.js';
import { lexStringLiteral } from '../expr/lexer.js';
import { ExpressionParser } from '../expr/parser.js';
import type { Source } from '../source.js';
import { isHorizontalSpace, skipHorizontalSpace } from './scanner.js';

const PATH_CHARS = /[A-Za-z0-9_./-]/;
const IDENT = /^[A-Za-z_][A-Za-z0-9_]*$/;

export interface PathToken {
  kind: 'path';
  value: string;
  start: number;
  end: number;
}

export interface IdentToken {
  kind: 'ident';
  value: string;
  start: number;
  end: number;
}

export type RawToken = PathToken | IdentToken;

export interface BlockBody {
  id: string | null;
  path: string | null;
  scope: ScopeItem[];
  end: number;
}

export class RawTagReader {
  constructor(
    private readonly source: Source,
    private readonly template: string,
    private readonly close: string,
    private readonly closeCount: number,
    private readonly fail: (code: 'E_PARSE_INVALID_PATH' | 'E_PARSE_INVALID_BLOCK_TAG' | 'E_PARSE_UNEXPECTED_TOKEN', start: number, end: number, message: string) => TemplateError,
  ) {}

  private get text(): string {
    return this.source.text;
  }

  private atClose(index: number): boolean {
    return this.text.startsWith(this.close.repeat(this.closeCount), index);
  }

  // Reads a quoted string or a run of path characters. Returns null at the close delimiter or end of input.
  private readToken(index: number): RawToken | null {
    index = skipHorizontalSpace(this.text, index);
    if (index >= this.text.length || this.atClose(index)) return null;
    const char = this.text[index] as string;
    if (char === '"' || char === "'") {
      const literal = lexStringLiteral(this.source, index, this.template);
      return { kind: 'path', value: literal.decoded, start: index, end: literal.end };
    }
    let end = index;
    while (end < this.text.length && PATH_CHARS.test(this.text[end] as string)) end++;
    const value = this.text.slice(index, end);
    if (value.includes('.') || value.includes('/')) return { kind: 'path', value, start: index, end };
    if (IDENT.test(value)) return { kind: 'ident', value, start: index, end };
    return { kind: 'ident', value: '', start: index, end: index + 1 };
  }

  // GRM-12: `+ path`.
  readIncludePath(index: number): { path: string; end: number } {
    const token = this.readToken(index);
    if (token === null || token.value === '') {
      const at = skipHorizontalSpace(this.text, index);
      throw this.fail('E_PARSE_INVALID_PATH', at, at + 1, 'include requires a path');
    }
    if (token.kind !== 'path') throw this.fail('E_PARSE_INVALID_PATH', token.start, token.end, `${JSON.stringify(token.value)} is not a path`);
    return { path: token.value, end: token.end };
  }

  // GRM-13 to GRM-15: `# [id] [path] {scope_item}`.
  readBlockBody(index: number): BlockBody {
    let id: string | null = null;
    let path: string | null = null;
    const scope: ScopeItem[] = [];
    let cursor = index;
    let token = this.readToken(cursor);
    if (token === null || token.value === '') {
      const at = skipHorizontalSpace(this.text, cursor);
      throw this.fail('E_PARSE_INVALID_BLOCK_TAG', at, at + 1, 'block tag requires an identifier or a path');
    }
    if (token.kind === 'path') {
      path = token.value;
      cursor = token.end;
    } else {
      id = token.value;
      cursor = token.end;
      token = this.readToken(cursor);
      if (token !== null && token.kind === 'path') {
        path = token.value;
        cursor = token.end;
      }
    }
    for (;;) {
      token = this.readToken(cursor);
      if (token === null) break;
      if (token.kind !== 'ident' || token.value === '') {
        throw this.fail('E_PARSE_INVALID_BLOCK_TAG', token.start, token.end, `unexpected ${JSON.stringify(this.text.slice(token.start, token.end))} in block tag`);
      }
      cursor = token.end;
      if (this.text[cursor] === ':') {
        const valueStart = cursor + 1;
        if (valueStart >= this.text.length || isHorizontalSpace(this.text.charCodeAt(valueStart))) {
          throw this.fail('E_PARSE_INVALID_BLOCK_TAG', cursor, cursor + 1, 'scope item requires a value after ":"');
        }
        const parser = new ExpressionParser(this.source, valueStart, { close: this.close, closeCount: this.closeCount, openIndex: null }, this.template);
        const expr: Expr = parser.parsePostfix(true);
        scope.push({ name: token.value, expr });
        cursor = parser.end;
      } else {
        scope.push({
          name: token.value,
          expr: { type: 'Var', name: token.value, span: this.source.span(token.start, token.end) },
        });
      }
    }
    return { id, path, scope, end: cursor };
  }
}
